draft-ietf-httpbis-client-cert-field.md

```diff
@@ -121,35 +121,73 @@ Finished messages.
+to generate an appropriate item.
 
 ## Client-Cert HTTP Header Field {#header}
 
-In the context of a TLS terminating reverse proxy (TTRP) deployment, the TTRP
+In the context of a TLS terminating reverse proxy (TTRP) deployment, the proxy
 makes the TLS client certificate available to the backend application with the
-following header field.
+Client-Cert HTTP header field. This field contains the end-entity certificate
+used by the client in the TLS handshake.
+
+Client-Cert is an Item Structured Header {{!RFC8941}}.  Its value MUST be a
+Byte Sequence ({{Section 3.3.5 of RFC8941}}).  Its ABNF is:
+
+~~~
+ Client-Cert = sf-binary
+~~~
+
+The value of the header is encoded as described in {{encoding}}.
+
+The `Client-Cert` header field is only for use in HTTP requests and MUST NOT be
+used in HTTP responses.  It is a single HTTP header field value as defined in
+Section 3.2 of {{?RFC7230}}, which MUST NOT have a list of values or occur
+multiple times in a request.
+
+## Client-Cert-Chain HTTP Header Field {#chain-header}
 
-Client-Cert:
-: The end-entity client certificate as an `EncodedCertificate` value.
+In the context of a TLS terminating reverse proxy (TTRP) deployment, the proxy
+MAY make the certificate chain used for validation of the end-entity certificate
+available to the backend application with the Client-Cert-Chain HTTP header
+field. This field contains certificates used by the proxy to validate the
+certificate used by the client in the TLS handshake. These certificates might or
+might not have been provided by the client during the TLS handshake.
 
-The `Client-Cert` header field defined herein is only for use in HTTP requests
-and MUST NOT be used in HTTP responses.  It is a single HTTP header field value
-as defined in Section 3.2 of {{?RFC7230}}, which MUST NOT have a list of values or
-occur multiple times in a request.
+Client-Cert-Chain is a List Structured Header {{!RFC8941}}.  Each item in the
+list MUST be a Byte Sequence ({{Section 3.3.5 of RFC8941}}) encoded as described
+in {{encoding}}.
+
+The header's ABNF is:
+
+~~~
+ Client-Cert-Chain = sf-list
+~~~
+
+The `Client-Cert-Chain` header field is only for use in HTTP requests and MUST
+NOT be used in HTTP responses.  It MAY have a list of values or occur multiple
+times in a request.  For header compression purposes, it might be advantageous
+to split lists into multiple instances.
+
+The first certificate in the list SHOULD directly certify the end-entity
+certificate provided in the `Client-Cert` header; each following certificate
+SHOULD directly certify the one immediately preceding it.  Because certificate
+validation requires that trust anchors be distributed independently, a
+certificate that specifies a trust anchor MAY be omitted from the chain,
+provided that the server is known to possess any omitted certificates.
+
+However, for maximum compatibility, servers SHOULD be prepared to handle
+potentially extraneous certificates and arbitrary orderings.
```

Comment:
Arbitrary ordering worries me here -- I think we can have a well-defined _order_ without necessarily having it be a complete (as in, no-gaps) certificate chain with specific endpoints. Maybe something like:

> The certificates in the `Cient-Cert-Chain` header field MUST be increasingly general, with the later certificates signing earlier certificates in the list.

... Or something along those lines. I think the warning about extraneous (and missing) certificates is reasonable, but I think the order should be predictable.

Reply:
Omitting the trust anchor, for the reasons mentioned, seems like a very reasonable thing. But I don't really see reason or value in allowing for gaps elsewhere, extraneous unrelated certificates, or unpredictable ordering. I plan to take editing pass over this section and try to tighten it up as such.

Reply:
I took the text here almost verbatim from TLS 1.3.  It recommends the proper ordering, but does not require it.  It certainly could be that they have constraints we don't -- legacy code being one, obviously -- but I'd want to have a concrete reason to diverge from what TLS uses in transferring the same thing.

Reply:
I am blissfully ignorant of the constraints and objectives that went into that bit of TLS 1.3 but, while the same thing is being transferred, the context of use is quite different. In TLS it's part of the elaborate authentication ceremony during the establishment of a connection.  With the Cient-Cert-Chain header, it's conveying the information post-authentication to another entity on every request on the established connection. My reasoning is more: why wouldn't we have the TTRP convey that information in a defined and predictable way?

Reply:
I agree with @b---c here, if this spec can be tighter in its generation requirements, I think it should be. Reaching out to the TLS community is not a bad idea though.

Reply:
This is the result of the way in which certification paths are constructed in practice, with co-signing and other things causing the ordering to be non-obvious.

Clearly, the order in which certificates appear can be linear, at least in the sense that the certificates can form a tree rooted on the EE certificate.  The problem is that it often needs to be a tree and not a chain because of the way in which different trust anchors accumulate in different endpoints.

This is far less likely to be a problem with client certificates as servers tend to carry the compatibility burdens, but I can't imagine that client certificates are completely immune to those sorts of compatibility problems either.

More fundamentally though: if the list of certificates that come out of TLS is not validated by the intermediary that inserts the header field, why do you think we can specify validation rules?  That is, if the certification path is constructed and validated by the back end system, the header field might be inserted with no knowledge of its contents. The only rule that makes any sense here is that the list of certificates match what the client included in its TLS handshake (or post-handshake authentication if you want to go there).

Reply:
My expectation of what Cient-Cert-Chain header is supposed to carry, which I think the text (largely from Mike) in the PR is in line with and is consistent with the thrust of the document in general, it that it's the chain/path that was used by the intermediary to validate the client's EE certificate. The intermediary that inserts the header field does the validation during the TLS handshake and the cert path that was constructed during the handshake is the chain that is conveyed to the back end.

Reply:
I don't think that you can assume that the intermediary will be performing path construction and validation, nor can you assume that the back-end server trusts it to do that.  While the backend might need to trust that the intermediary validates the in-TLS signature relative to the EE certificate, this design allows the backend to perform all the path construction independently.

If the back-end has different trust anchors then the intermediary, it will want to have all the information presented in the TLS handshake.

Reply:
Honestly, I'm at a bit of a loss as to how to respond (and have spent way more time than I'd like trying to do so). That is exactly the assumption I want to make and what I view as an objective of the document to facilitate - where the proxy does the validation work and only exposes the information to back end. I view separating out the validation responsibilities like that as being rife with potential for subtle unexpected problems.  But maybe I'm in the weeds on this. I was rather skeptical of the purported need to expose more than the client's EE cert in the first place. Yet a number of folks expressed a need/want for it. And here we are. 

I also tried to look at docs of a few proxies and TLS implementations to get an idea of existing practice and/or what's even generally possible. And, to me anyway, everything is pretty ambiguous about what is exposed - what the client provided or what the server constructed. I suppose that if one reads the text of this PR without the bias I came with, it's consistent with that. 

All of this rambling is to say that l relent and I'll put that sentence back as well as the SHOULDs in the paragraph prior.

Reply:
b251935ece5105c3b98aa078610e0fdeed9581bd puts that sentence back in the text as well as the SHOULDs into the paragraph prior.